main/core/injector.ts

```diff
@@ -191,56 +262,56 @@ export class Injector implements IInjector {
         ancestry: any[] = [],
         options?: IConstructionOptions<T>,
     ): InstanceType<T> {
+        if (this._isDestroyed) {
+            throw new Error(
+                `Invalid operation, the current injector instance is marked as destroyed. Injector Id: [${this.id}]`,
+            );
+        }
+
         // Measuring time taken
         const start = Date.now();
+        let instance: any;
 
         const constructorType: any =
-            typeof typeOrToken === 'string' ? this.tokenCache.getTypeForToken(typeOrToken) : typeOrToken;
+            typeof typeOrToken === 'string' ? this.getTypeForToken(this, typeOrToken) : typeOrToken;
 
         if (constructorType === undefined) {
             throw new Error(
                 `Cannot resolve Type with token '${typeOrToken}' as no types have been registered against that token value`,
             );
         }
 
-        // We use a special ID to determine a type could be an injector type
-        const isInjectorType = constructorType.typeId === Injector.typeId;
-
-        // Lets check the cache and see if we have an instance ready to service
-        let instance = this.cache.resolve(constructorType);
-
-        // Support resolving Injector as injectable if found in the dependency tree
-        if (isInjectorType) {
-            if (instance == null) {
-                // Does the instance being requested share the same prototype as our current injector instance (Or it has no create function)
-                if (
-                    constructorType === Injector ||
-                    (constructorType !== Injector && constructorType.create === undefined)
-                ) {
-                    instance = this;
-                    this.cache.update(constructorType, instance);
-                } else {
-                    /* If we are here then we have multiple injector versions in play.
-                     * we check to see if the type has the static create method.  If
-                     * we find it we can instance it using that method. */
-                    instance = constructorType.create();
-                    this.cache.update(constructorType, instance);
-                }
-            }
-        }
+        ancestry.push(constructorType);
+
+        // If injector type then return `this` if not try and resolve from cache
+        const isInjectorType = constructorType === Injector || constructorType.typeId === INJECTOR_TYPE_ID;
+        instance = isInjectorType ? this : this.cache.resolve(constructorType);
 
         if (instance == null) {
-            // If an external resolution strategy has been set, delegate all responsibility to it
+            // Lets see if the type has an associated registration. If root and answer is no then bail
+            const registration = this._registrations.get(constructorType);
```

Comment:
are there tests for this stuff? Doesn't look like any tests have been modified (although maybe I've just not got to them yet)

Reply:
Yes tests have range of scenarios with full depth testing.